File: hellospring/src/main/java/kopher/hellospring/exrate/WebApiExRateProvider.java
package kopher.hellospring.exrate;

import kopher.hellospring.payment.ExRateProvider;
import org.springframework.stereotype.Component;
import tools.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigDecimal;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.util.stream.Collectors;

@Component
public class WebApiExRateProvider implements ExRateProvider {
    @Override
    public BigDecimal getExRate(String currency) throws IOException {
        URL url = URI.create("https://open.er-api.com/v6/latest/" + currency).toURL();
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        BufferedReader br = new BufferedReader(new InputStreamReader(connection.getInputStream()));

        String response = br.lines().collect(Collectors.joining());
        br.close();

        System.out.println("response = " + response);
        ObjectMapper mapper = new ObjectMapper();
        ExRateData data = mapper.readValue(response, ExRateData.class);

        System.out.println("data.rates().get(\"KRW\") = " + data.rates().get("KRW"));

        return data.rates().get("KRW");
    }
}

File: spring-principle/src/main/kotlin/jp/kopher/springprinciple/WebApiExRateProvider.kt
package jp.kopher.springprinciple

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.io.BufferedReader
import java.io.InputStreamReader
import java.math.BigDecimal
import java.net.URL
import java.util.stream.Collectors

class WebApiExRateProvider {
    fun getWebExRate(currency: String): BigDecimal {
        val url = URL(" https://open.er-api.com/v6/latest/${currency}")
        val connection = url.openConnection()
        val br = BufferedReader(InputStreamReader(connection.getInputStream(), "UTF-8"))
        val response = br.lines().collect(Collectors.joining())
        br.close()

        val mapper = jacksonObjectMapper()
        val data = mapper.readValue(response, ExRateData::class.java)
        return data.rates["KRW"] ?: BigDecimal.ZERO
    }
}